clang/Token.d

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright: Copyright (c) 2011 Jacob Carlborg. All rights reserved.
+ * Authors: Jacob Carlborg
+ * Version: Initial created: Jan 1, 2012
+ * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0)
+ */
+module clang.Token;
+
+import std.typecons;
+import std.conv : to;
+import clang.c.Index;
+import clang.SourceLocation;
+import clang.SourceRange;
+import clang.Type;
+import clang.Util;
+import clang.Visitor;
+import clang.Cursor;
+
+struct Token
+{
+    private struct Container
+    {
+        CXTranslationUnit translationUnit;
+        CXToken* tokens;
+        ulong numTokens;
+
+        ~this() 
+        {
+            if (tokens != null)
+            {
+                clang_disposeTokens(
+                    translationUnit, 
+                    tokens, 
+                    to!uint(numTokens));
+            }
+        }
+    }
+
+    private const RefCounted!Container _container;
+    size_t _index;
+
+    @property
+    private Container* _containerPtr() const
+    {
+        return cast(Container*) &(_container.refCountedPayload());
+    }
+
+    @property static Cursor empty ()
+    {
+        auto r = clang_getNullCursor();
+        return Cursor(r);
+    }
+
+    @property 
+    string spelling() const
+    {
+        return toD(clang_getTokenSpelling(
+            _containerPtr.translationUnit, 
+            _containerPtr.tokens[_index]));
+    }
+
+    @property
+    CXTokenKind kind() const 
+    {
+        return clang_getTokenKind(_containerPtr.tokens[_index]);
+    }
+
+    @property
+    SourceRange extent() const
+    {
+        return SourceRange(
+            clang_getTokenExtent(
+                _containerPtr.translationUnit, 
+                _containerPtr.tokens[_index]));
+    }
+
+    @property
+    string toString() const 
+    {
+        import std.format: format;
+        import std.conv: to;
+        return format("Token(%s, %s)", to!string(kind), spelling);
+    }
+}
+
+struct TokenRange
+{
+    private const RefCounted!(Token.Container) _container;
+    private size_t _begin;
```

Comment:
Eh, I don't like this because '_' simplifies some things - e.g. you do not have name collisions with function arguments (common case in constructor). But, ok...